Adding snap integration

knkski opened the pull request:
Allows installing lemmy via `sudo snap install lemmy`.

knkski commented on server/lemmy_utils/src/settings.rs:
```diff
@@ -123,7 +123,11 @@ impl Settings {
 
   pub fn get_config_defaults_location() -> String {
-    env::var("LEMMY_CONFIG_LOCATION").unwrap_or_else(|_| CONFIG_FILE_DEFAULTS.to_string())
+    env::var("LEMMY_CONFIG_DEFAULTS_LOCATION").unwrap_or_else(|_| CONFIG_FILE_DEFAULTS.to_string())
```
I renamed this env var, which is backwards-incompatible. It seemed confusing though to have `CONFIG_FILE` not be equivalent to `LEMMY_CONFIG_LOCATION`. Is there a better name that I should use down below for the equivalent of `CONFIG_FILE`?

dessalines replied:
That change seems good to me. Make sure you grep the codebase when making changes tho, I found an instance of this env var in the docs at `docs/src/administration_configuration.md`

dessalines reviewed (changes requested):
One more thing, you'll need to add a section to the installation docs for how to use / install this snap.

I'm good with merging this after those changes, but just note that the main devs don't run ubuntu, so we can't help anyone debug this; we'll have to direct any questions / problems with it to you. We've had people add different install methods (like kubernetes), then ghost after a few weeks, leaving them unmaintained. 

We're also moving to an isomorphic app setup in a few days, and that will likely mean some of this will need to be changed/updated after that's done.

knkski commented:
Thanks for the review. I've updated the PR to include docs that cover installation and configuration via snap. I also updated the existing reference to `LEMMY_CONFIG_LOCATION` in `administration_configuration.md ` to cover both of the new environment variables.

dessalines commented:
Bump, there are some conflicts here.

BTW I missed this bc you used force-push, which doesn't work with changes requested.

knkski commented:
Ah sorry, will avoid force pushes in the future. Working on updating this with the latest changes, will push up when ready.

knkski commented:
@dessalines: Apologies for force pushing again, but I had to rebase with the changes introduced by the isomorphic app work. I don't think there's a way of pushing up all those rebased commits without force pushing, but I also re-requested a review from you.

knkski commented on lemmy_utils/src/request.rs:
```diff
@@ -58,7 +58,7 @@ pub(crate) async fn fetch_iframely(
   url: &str,
 ) -> Result<IframelyResponse, LemmyError> {
-  let fetch_url = format!("http://iframely/oembed?url={}", url);
+  let fetch_url = format!("http://localhost:8061/oembed?url={}", url);
```
What should this be set to? There isn't an `iframely_url` option in the Settings struct, should one be added?

Nutomic replied:
Yep this will require a new setting.

knkski commented on lemmy_utils/src/request.rs:
```diff
@@ -88,7 +88,7 @@ pub(crate) async fn fetch_pictrs(
 
   let fetch_url = format!(
-    "http://pictrs:8080/image/download?url={}",
+    "http://localhost:1235/image/download?url={}",
```
I assume that this should be pointed to `Settings.pictrs_url`, but wanted to make sure before switching it.

Nutomic replied:
You're right, looks like we forgot to adjust that.

knkski commented on src/main.rs:
```diff
@@ -100,7 +100,7 @@ async fn main() -> Result<(), LemmyError> {
       .configure(nodeinfo::config)
       .configure(webfinger::config)
-      .service(actix_files::Files::new("/docs", "/app/documentation"))
+      .service(actix_files::Files::new("/docs", "/snap/lemmy/current/ui/dist/documentation"))
```
Should this be added as a configuration option to the `Settings` struct?

Nutomic replied:
Most likely, yes.

Nutomic commented on CODEOWNERS:
```diff
@@ -0,0 +1,2 @@
+*       @dessalines
```
Please add me too in this line :)

Nutomic commented on snap/hooks/configure:
```diff
@@ -0,0 +1,82 @@
+# End config
+echo '}' >> $SNAP_DATA/config.hjson
+
+snapctl restart ${SNAP_NAME}.lemmy
```
This script looks pretty hard to maintain in case we add or change any settings. Maybe you could use `jq` to simplify it?

knkski replied:
Good idea. I was toying with including Python, which would be a pretty huge dependency for just some JSON munging. I'll take a look at `jq` though, that should be small enough to fit in the snap quite reasonably :+1:

Nutomic commented:
As a more general discussion, how do you envision the maintenance of the Lemmy snap going? It looks like we need to push new versions manually to snapcraft.io, are you going to take care of that? And are you going to adjust it to any changes in the config or dependencies? Because @dessalines and me are already busy enough.

To be honest, I think it would be better to put this into a separate git repo (which could be in the LemmyNet org).

knkski commented:
@Nutomic: I was planning on maintaining the snap, and updating it based on config changes. The advantage to including it in this repository as opposed a separate one is that snapcraft.io actually has integration with github, and can build the snap and release it whenever a commit is merged to master.

dessalines commented:
@knkski Let us know if you need help externalizing these urls / dirs : https://github.com/LemmyNet/lemmy/blob/main/lemmy_utils/src/settings.rs#L9

dessalines commented on docs/src/administration_configuration.md:
```diff
@@ -5,7 +5,12 @@ The configuration is based on the file
 can copy the options you want to change into your local `config.hjson` file.
 
-To use a different `config.hjson` location than the current directory, set the environment variable `LEMMY_CONFIG_LOCATION`. Make sure you copy the `defaults.hjson` if you do this, otherwise you will be missing settings.
+The `defaults.hjson` and `config.hjson` files are located at `config/defaults.hjson` and
+`config/config.hjson`, respectively. To change these default locations, you can set these
+two environment variables:
+
+    LEMMY_CONFIG_LOCATION           # config.hjson
+    LEMMY_CONFIG_DEFAULTS_LOCATION  # defaults.hjson
```
There's some weird line breaking and indentation formatting here. This is a markdown file so that's not necessary.

knkski replied:
I added the indentation because it's a markdown file, I figured it would be good to have those environment variables visually called out in the documentation when someone's scanning through it. I can change that to something else though if you'd like.

As far as the line breaking, my default line length when typing is apparently longer than Github's max line length for word wrapping, hence the weird wrapping. At any rate, just removed the manual line breaks, since you're right that it's going to be rendered via markdown anyways.

dessalines commented on docs/src/administration_configuration.md:
```diff
@@ -22,3 +27,22 @@ following commands:
+
+### Snap Configuration
+
+If you have installed Lemmy as a snap, configuration can be done via `snap set`. Each property
+in the `config.hjson` has a corresponding `snap set` parameter, with nested properties
+separated by dots. For example, to set the hostname of your Lemmy instance, run
+
+    sudo snap set lemmy hostname=mylemmy
+
+To set the admin username before setup, run
+
+    sudo snap set lemmy setup.admin-username=admin
+
+Note that each property name that contains an underscore in `config.hjson` must be set using
+a dash via `snap set`. The dash is replaced by an underscore in the generated `config.hjson`.
+For example, `sudo snap set lemmy setup.jwt-secret=foo` is equivalent to this
+`config.hjson`:
```
Same weird line breaking issues.

dessalines commented on docs/src/administration_install_snap.md:
```diff
@@ -0,0 +1,21 @@
+# Snap Installation
```
Same.

dessalines commented on rust-toolchain:
```diff
@@ -0,0 +1 @@
+1.45.0
```
Hrm, I'm not sure about this one, we generally have to use nightly for our dev environments. @Nutomic ?

Nutomic replied:
Well `cargo +nightly check` is still working fine with that file present. But I'm not sure why we should specify a particular Rust version at all, as it should always work with the latest stable version (unless they break things again).

dessalines replied:
We do see that `config-rs` is broken on **current** nightly, so it might be good to have this, but have it match our docker. Because nightly is required for the cargo formatting and clippy stuff.

knkski replied:
I added this file due to this error while compiling with 1.46.0: https://github.com/rust-lang/futures-rs/issues/2202

I haven't tried it again after rebasing with the isomorphic app changes. I'll try it again without this file, and remove this file if compilation succeeds. In general, if we're held back from using the latest stable version of Rust, is this method the one we'd want to go with?

dessalines replied:
I gotcha, no its fine to leave it in, we've actually been having some problems with various rust versions too, and this would be a good way to enforce a working version for devs.

Nutomic commented on snap/hooks/install:
```diff
@@ -0,0 +1,41 @@
+    difficulty: medium
+  }
+}
+EOF
```
It would be good if you could include this file directly from `config/defaults.hjson`, so we dont have to add new variables in two places. Values that need to be adjusted, like jwt_secret or docs_dir, can go into `config.hjson` or into an env var.

Nutomic commented on snap/local/conf/nginx.conf:
```diff
@@ -0,0 +1,52 @@
+            proxy_pass http://localhost:8540/api/v1;
+            proxy_http_version 1.1;
+            proxy_set_header Upgrade $http_upgrade;
+            proxy_set_header Connection "upgrade";
```
This one needs `x-real-ip` etc headers (see #1215).

dessalines commented:
We talked about this, and because we've had people that have ghosted on unofficial install methods before (and have had to remove them), we're reticent to add potentially dozens of unofficial install methods, that we have no idea how to do support for, into the main lemmy repo. What will work: 

- Create your own repo, maybe called `lemmy-snap`, or whatever you like.
- If you need the lemmy repo, add it as a git submodule to yours.
- Put your install documentation in your `README.md`.
- We'll add a section in our docs for "unofficial installation methods", and link to your snap repo.
- Alter this PR to just be the config changes required for your snap install.

Nutomic commented:
We can also put your `lemmy-snap` repo into the LemmyNet organization if you want.

knkski commented:
@dessalines: Understood. I'll close this PR and break up the config changes into separate PRs.

@Nutomic: Sure, that would be great

dessalines commented:
@Nutomic Could you create that repo, then give @knkski admin on it.

Nutomic commented:
Done, you should have an invite @knkski
